src/main/java/gregtech/api/unification/stack/ItemVariantMap.java

```diff
@@ -0,0 +1,153 @@
+package gregtech.api.unification.stack;
+
+import gregtech.api.GTValues;
+import gregtech.api.util.GTUtility;
+import net.minecraft.item.ItemStack;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Set;
+
+/**
+ * An abstraction of dictionary-like collection with each item variant as keys.
+ * <p>
+ * Despite the name, the actual implementation of the class may be unrelated to maps.
+ * Nonetheless, this class supports map-like access to whatever form of data they're
+ * holding, with additional methods like {@link #hasNonWildcardEntry()} to help
+ * certain operations special to item variants.
+ *
+ * @param <E> type of the elements
+ * @see SingleItemVariantMap
+ * @see MultiItemVariantMap
+ */
+public interface ItemVariantMap<E> {
+
+    /**
+     * @return {@code true} if there's any nonnull value associated with some item
+     * metadata, excluding metadata value of {@code 32767}; i.e. "wildcard".
```

Comment:
Might be better to `{@link}` to `GTValues.W` or the constant in `OreDictionary` here.